### DISCUSSION TITLE
Add prCreation option 'stable' to create PRs after stability days

### DISCUSSION THREAD

**Discussion by bigpopakap** [renovatebot/renovate#15764] [Request Help] [answered]
### What would you like Renovate to be able to do?

With my current config (a simplified example)
```
{
    extends: ["config:base"],

    rebaseWhen: "behind-base-branch",
    automerge: true,
 
    // Avoids a bunch of unnecessary rebases (and therefore CI usage)
    // This situations is described in #4806, but it seems difficult to update Renovate to merge/rebase smartly, so I'm just manually limiting to 1 PR at a time
    prConcurrentLimit: 1,
 
    packageRules: [
        {
          // Major dependencies
          matchDepTypes: ["dependencies"],
          matchUpdateTypes: ["major"],
    
          // Wait a good long number of days before creating the PRs
          stabilityDays: 5,
          prCreation: "not-pending",
        }
    ]
}
```

I'd like for all non-major version changes to be automerged immediately (assuming they pass CI)
While a major version is waiting for `5` stability days, all non-major version changes should continue to be automerged immediately, without being blocked behind the major version PR

It seems, however, that it works like this:
1. a new major version comes in
1. renovate creates the branch for that major version
1. CI starts running on that branch
1. CI completes, and the PR is created (because the stability check doesn't count as a "pending" status? I see that #7800 is not completed though so I'm confused here)
1. Now the major version's PR waits for 5 days to automerge because of the stability check
1. In this time, no other versions are able to be automerged, because of `prConcurrentLimit: 1`

### If you have any ideas on how this should be implemented, please tell us here.

#4806 is at the heart of this request since I could omit `prConcurrentLimit: 1`. This solution would be ideal because it could maybe continue automerging new updates even if one of them fails CI, which is a limitation of my `prConcurrentLimit: 1` approach to avoiding unnecessary rebases. However, it seems like this is a difficult problem to solve

I imagine this might be a new option, like `prCreation: 'stable'`. With `prCreation: 'stable'`, the PRs will be created once the stability check has passed. Other status checks will still be run, but only the stability check will block PR creation.

I'd be happy to work on a solution, but I would need some pretty clear guidance pointing me to the right pieces of code to update.

### Is this a feature you are interested in implementing yourself?

Maybe

**Comment by rarkins** [comment]
1. prConcurrentLimit=1 is going to cause you problems no matter what you do, because it's going to stop other PRs whenever there's *any* reason the 1 PR can't be merged immediately
2. I think your problem is mostly solved with https://docs.renovatebot.com/configuration-options/#internalchecksfilter and setting to "strict"

**Reply by bigpopakap** [reply]
Ok, interesting. Thanks for converting this to a discussion, and for the tip! I'm going to try it out, but it may be a couple days (watching how PRs are created and merged) before I can figure out whether that does what I want.

If not, I'll be sure to formulate my request a little more clearly with updated information. Even if this suggestion doesn't work, it'll help me make my request more clear, or uncover other problems with my config. Thank you!

(sorry for the noise... I haven't used Discussions before, and I didn't realize my responses were being written in separate threads. This comment is copied from something I wrote previously, just to get it in the right place)

**Comment by bigpopakap** [comment, marked as answer]
Ok I'm having a slightly different issue now. I think this issue is separate from everything we've discussed before. My first message was an attempt at explaining the situation, but I got one important thing wrong (I failed to take into account that CI only builds pull requests, not every branch).

The new situation:

I have set my CI to only build PRs, in order to save CI usage. This way, a developer can save in-progress work to a remote branch, but it won't waste CI resources until that work is ready and the developer creates a PR for it.

What happens now is:
1. a new major version comes in
1. renovate creates the branch for that major version
1. renovate waits for the `stabilityDays` check before creating the PR (because I have `stabilityDays=5` and `prCreation="not-pending"` for major version updates
1. after waiting the appropriate number of days for the stability check to pass (or if forcing creation of the PR manually), renovate creates the PR
1. renovate merges immediately 😢 because CI has not already been run on that branch, so renovate sees a single passed check for stability, and merges.

I like that Renovate will wait for stability days to pass before creating the PR. But then I'd like to to wait for CI to pass before auto-merging. This could be solved by having CI build every branch, but as mentioned before I'm trying to limit CI usage, particularly because of `rebaseWhen="behind-base-branch"`

<details>
<summary>Summing up the other solutions discussed</summary>

So for `prConcurrentLimit=1`, that limitation is ok with me. If CI fails and that causes my whole single-file line of PRs to be blocked, that's ok. I'll just fix CI when I can, and then everything can continue. Saving CI usage is most important to me, so I could either do `rebaseWhen="behind-base-branch" && prConcurrentLimit=1` or relax the `rebaseWhen` requirement.

I think [internalchecksfilter](https://docs.renovatebot.com/configuration-options/#internalchecksfilter) is for a separate issue. As I understand it, that concerns what happens for multiple versions of the same package (should it use the latest, or go one-by-one).
</details>

**Reply by rarkins** [reply]
There are existing issues for that:

https://github.com/renovatebot/renovate/issues/13483
https://github.com/renovatebot/renovate/issues/7800

**Reply by bigpopakap** [reply]
Oh yes, I think my answer for now is:
* Just make CI build for every branch, not just PRs
* Follow #7800, and once that's fixed I can update CI to go back to building only PRs
* Set `internalChecksFilter='strict'` to limit the number of PRs that will be created and possibly block other PRs (because of my choice to stick with `prConcurrentLimit = 1`)

That will basically leave me with the following config options, and I'll just monitor how I like that behavior: `rebaseWhen="behind-base-branch"` and `prConcurrentLimit=1` for everything, and `stabilityDays=5` and `prCreation="not-pending"` for major versions

Summing it up here in case someone else comes across this discussion with a similar situation

Thank you for the guidance! I'm in such awe at how awesome Renovate is. It's really a magical tool, saving so much manual time, and keeping so many software projects up to date and secure! 🏆
